File: trunk/cpp/DIS/UAPdu.cpp
#include <DIS/UAPdu.h> 

using namespace DIS;


UAPdu::UAPdu() : DistributedEmissionsPdu(),
   _emittingEntityID(), 
   _eventID(), 
   _stateChangeIndicator(0), 
   _pad(0), 
   _passiveParameterIndex(0), 
   _propulsionPlantConfiguration(0), 
   _numberOfShafts(0), 
   _numberOfAPAs(0), 
   _numberOfUAEmitterSystems(0)
{
    setPduType( 29 );
}

UAPdu::~UAPdu()
{
    _shaftRPMs.clear();
    _apaData.clear();
    _emitterSystems.clear();
}

EntityID& UAPdu::getEmittingEntityID() 
{
    return _emittingEntityID;
}

const EntityID& UAPdu::getEmittingEntityID() const
{
    return _emittingEntityID;
}

void UAPdu::setEmittingEntityID(const EntityID &pX)
{
    _emittingEntityID = pX;
}

EventID& UAPdu::getEventID() 
{
    return _eventID;
}

const EventID& UAPdu::getEventID() const
{
    return _eventID;
}

void UAPdu::setEventID(const EventID &pX)
{
    _eventID = pX;
}

char UAPdu::getStateChangeIndicator() const
{
    return _stateChangeIndicator;
}

void UAPdu::setStateChangeIndicator(char pX)
{
    _stateChangeIndicator = pX;
}

char UAPdu::getPad() const
{
    return _pad;
}

void UAPdu::setPad(char pX)
{
    _pad = pX;
}

unsigned short UAPdu::getPassiveParameterIndex() const
{
    return _passiveParameterIndex;
}

void UAPdu::setPassiveParameterIndex(unsigned short pX)
{
    _passiveParameterIndex = pX;
}

unsigned char UAPdu::getPropulsionPlantConfiguration() const
{
    return _propulsionPlantConfiguration;
}

void UAPdu::setPropulsionPlantConfiguration(unsigned char pX)
{
    _propulsionPlantConfiguration = pX;
}

unsigned char UAPdu::getNumberOfShafts() const
{
   return _shaftRPMs.size();
}

unsigned char UAPdu::getNumberOfAPAs() const
{
   return _apaData.size();
}

unsigned char UAPdu::getNumberOfUAEmitterSystems() const
{
   return _emitterSystems.size();
}

std::vector<ShaftRPMs>& UAPdu::getShaftRPMs() 
{
    return _shaftRPMs;
}

const std::vector<ShaftRPMs>& UAPdu::getShaftRPMs() const
{
    return _shaftRPMs;
}

void UAPdu::setShaftRPMs(const std::vector<ShaftRPMs>& pX)
{
     _shaftRPMs = pX;
}

std::vector<APAData>& UAPdu::getApaData() 
{
    return _apaData;
}

const std::vector<APAData>& UAPdu::getApaData() const
{
    return _apaData;
}

void UAPdu::setApaData(const std::vector<APAData>& pX)
{
     _apaData = pX;
}

std::vector<AcousticEmitterSystemData>& UAPdu::getEmitterSystems() 
{
    return _emitterSystems;
}

const std::vector<AcousticEmitterSystemData>& UAPdu::getEmitterSystems() const
{
    return _emitterSystems;
}

void UAPdu::setEmitterSystems(const std::vector<AcousticEmitterSystemData>& pX)
{
     _emitterSystems = pX;
}

void UAPdu::marshal(DataStream& dataStream) const
{
    DistributedEmissionsPdu::marshal(dataStream); // Marshal information in superclass first
    _emittingEntityID.marshal(dataStream);
    _eventID.marshal(dataStream);
    dataStream << _stateChangeIndicator;
    dataStream << _pad;
    dataStream << _passiveParameterIndex;
    dataStream << _propulsionPlantConfiguration;
    dataStream << ( unsigned char )_shaftRPMs.size();
    dataStream << ( unsigned char )_apaData.size();
    dataStream << ( unsigned char )_emitterSystems.size();

     for(size_t idx = 0; idx < _shaftRPMs.size(); idx++)
     {
        ShaftRPMs x = _shaftRPMs[idx];
        x.marshal(dataStream);
     }


     for(size_t idx = 0; idx < _apaData.size(); idx++)
     {
        APAData x = _apaData[idx];
        x.marshal(dataStream);
     }


     for(size_t idx = 0; idx < _emitterSystems.size(); idx++)
     {
        AcousticEmitterSystemData x = _emitterSystems[idx];
        x.marshal(dataStream);
     }

}

void UAPdu::unmarshal(DataStream& dataStream)
{
    DistributedEmissionsPdu::unmarshal(dataStream); // unmarshal information in superclass first
    _emittingEntityID.unmarshal(dataStream);
    _eventID.unmarshal(dataStream);
    dataStream >> _stateChangeIndicator;
    dataStream >> _pad;
    dataStream >> _passiveParameterIndex;
    dataStream >> _propulsionPlantConfiguration;
    dataStream >> _numberOfShafts;
    dataStream >> _numberOfAPAs;
    dataStream >> _numberOfUAEmitterSystems;

     _shaftRPMs.clear();
     for(size_t idx = 0; idx < _numberOfShafts; idx++)
     {
        ShaftRPMs x;
        x.unmarshal(dataStream);
        _shaftRPMs.push_back(x);
     }

     _apaData.clear();
     for(size_t idx = 0; idx < _numberOfAPAs; idx++)
     {
        APAData x;
        x.unmarshal(dataStream);
        _apaData.push_back(x);
     }

     _emitterSystems.clear();
     for(size_t idx = 0; idx < _numberOfUAEmitterSystems; idx++)
     {
        AcousticEmitterSystemData x;
        x.unmarshal(dataStream);
        _emitterSystems.push_back(x);
     }
}


bool UAPdu::operator ==(const UAPdu& rhs) const
 {
     bool ivarsEqual = true;

     ivarsEqual = DistributedEmissionsPdu::operator==(rhs);

     if( ! (_emittingEntityID == rhs._emittingEntityID) ) ivarsEqual = false;
     if( ! (_eventID == rhs._eventID) ) ivarsEqual = false;
     if( ! (_stateChangeIndicator == rhs._stateChangeIndicator) ) ivarsEqual = false;
     if( ! (_pad == rhs._pad) ) ivarsEqual = false;
     if( ! (_passiveParameterIndex == rhs._passiveParameterIndex) ) ivarsEqual = false;
     if( ! (_propulsionPlantConfiguration == rhs._propulsionPlantConfiguration) ) ivarsEqual = false;

     for(size_t idx = 0; idx < _shaftRPMs.size(); idx++)
     {
        if( ! ( _shaftRPMs[idx] == rhs._shaftRPMs[idx]) ) ivarsEqual = false;
     }


     for(size_t idx = 0; idx < _apaData.size(); idx++)
     {
        if( ! ( _apaData[idx] == rhs._apaData[idx]) ) ivarsEqual = false;
     }


     for(size_t idx = 0; idx < _emitterSystems.size(); idx++)
     {
        if( ! ( _emitterSystems[idx] == rhs._emitterSystems[idx]) ) ivarsEqual = false;
     }


    return ivarsEqual;
 }

int UAPdu::getMarshalledSize() const
{
   int marshalSize = 0;

   marshalSize = DistributedEmissionsPdu::getMarshalledSize();
   marshalSize = marshalSize + _emittingEntityID.getMarshalledSize();  // _emittingEntityID
   marshalSize = marshalSize + _eventID.getMarshalledSize();  // _eventID
   marshalSize = marshalSize + 1;  // _stateChangeIndicator
   marshalSize = marshalSize + 1;  // _pad
   marshalSize = marshalSize + 2;  // _passiveParameterIndex
   marshalSize = marshalSize + 1;  // _propulsionPlantConfiguration
   marshalSize = marshalSize + 1;  // _numberOfShafts
   marshalSize = marshalSize + 1;  // _numberOfAPAs
   marshalSize = marshalSize + 1;  // _numberOfUAEmitterSystems

   for(int idx=0; idx < _shaftRPMs.size(); idx++)
   {
        ShaftRPMs listElement = _shaftRPMs[idx];
        marshalSize = marshalSize + listElement.getMarshalledSize();
    }


   for(int idx=0; idx < _apaData.size(); idx++)
   {
        APAData listElement = _apaData[idx];
        marshalSize = marshalSize + listElement.getMarshalledSize();
    }


   for(int idx=0; idx < _emitterSystems.size(); idx++)
   {
        AcousticEmitterSystemData listElement = _emitterSystems[idx];
        marshalSize = marshalSize + listElement.getMarshalledSize();
    }

    return marshalSize;
}

File: trunk/open-dis/java/edu/nps/moves/dis/UaPdu.java
package edu.nps.moves.dis;

import java.util.*;
import java.io.*;
import javax.xml.bind.annotation.*;

/**
 * Section 5.3.7.3. Information about underwater acoustic emmissions. This requires manual cleanup.  The beam data records should ALL be a the finish, rather than attached to each emitter system. UNFINISHED
 *
 * Copyright (c) 2008, MOVES Institute, Naval Postgraduate School. All rights reserved.
 * This work is licensed under the BSD open source license, available at https://www.movesinstitute.org/licenses/bsd.html
 *
 * @author DMcG
 * @version $Id:$
 */
public class UaPdu extends DistributedEmissionsFamilyPdu implements Serializable {

    /** ID of the entity that is the source of the emission */
    protected EntityID emittingEntityID = new EntityID();
    /** ID of event */
    protected EventID eventID = new EventID();
    /** This field shall be used to indicate whether the data in the UA PDU represent a state update or data that have changed since issuance of the last UA PDU */
    protected byte stateChangeIndicator;
    /** padding */
    protected byte pad;
    /** This field indicates which database record (or file) shall be used in the definition of passive signature (unintentional) emissions of the entity. The indicated database record (or  file) shall define all noise generated as a function of propulsion plant configurations and associated  auxiliaries. */
    protected int passiveParameterIndex;
    /** This field shall specify the entity propulsion plant configuration. This field is used to determine the passive signature characteristics of an entity. */
    protected short propulsionPlantConfiguration;
    /**  This field shall represent the number of shafts on a platform */
    protected short numberOfShafts;
    /** This field shall indicate the number of APAs described in the current UA PDU */
    protected short numberOfAPAs;
    /** This field shall specify the number of UA emitter systems being described in the current UA PDU */
    protected short numberOfUAEmitterSystems;
    /** shaft RPM values */
    protected List<ShaftRPMs> shaftRPMs = new ArrayList<ShaftRPMs>();
    /** apaData */
    protected List<ApaData> apaData = new ArrayList<ApaData>();
    protected List<AcousticEmitterSystemData> emitterSystems = new ArrayList<AcousticEmitterSystemData>();

    /** Constructor */
    public UaPdu() {
        setPduType((short) 29);
    }

    @Override
    public int getMarshalledSize() {
        int marshalSize = 0;

        marshalSize = super.getMarshalledSize();
        marshalSize = marshalSize + emittingEntityID.getMarshalledSize();  // emittingEntityID
        marshalSize = marshalSize + eventID.getMarshalledSize();  // eventID
        marshalSize = marshalSize + 1;  // stateChangeIndicator
        marshalSize = marshalSize + 1;  // pad
        marshalSize = marshalSize + 2;  // passiveParameterIndex
        marshalSize = marshalSize + 1;  // propulsionPlantConfiguration
        marshalSize = marshalSize + 1;  // numberOfShafts
        marshalSize = marshalSize + 1;  // numberOfAPAs
        marshalSize = marshalSize + 1;  // numberOfUAEmitterSystems
        for (int idx = 0; idx < shaftRPMs.size(); idx++) {
            ShaftRPMs listElement = shaftRPMs.get(idx);
            marshalSize = marshalSize + listElement.getMarshalledSize();
        }
        for (int idx = 0; idx < apaData.size(); idx++) {
            ApaData listElement = apaData.get(idx);
            marshalSize = marshalSize + listElement.getMarshalledSize();
        }
        for (int idx = 0; idx < emitterSystems.size(); idx++) {
            AcousticEmitterSystemData listElement = emitterSystems.get(idx);
            marshalSize = marshalSize + listElement.getMarshalledSize();
        }

        return marshalSize;
    }

    public void setEmittingEntityID(EntityID pEmittingEntityID) {
        emittingEntityID = pEmittingEntityID;
    }

    @XmlElement
    public EntityID getEmittingEntityID() {
        return emittingEntityID;
    }

    public void setEventID(EventID pEventID) {
        eventID = pEventID;
    }

    @XmlElement
    public EventID getEventID() {
        return eventID;
    }

    public void setStateChangeIndicator(byte pStateChangeIndicator) {
        stateChangeIndicator = pStateChangeIndicator;
    }

    @XmlAttribute
    public byte getStateChangeIndicator() {
        return stateChangeIndicator;
    }

    public void setPad(byte pPad) {
        pad = pPad;
    }

    @XmlAttribute
    public byte getPad() {
        return pad;
    }

    public void setPassiveParameterIndex(int pPassiveParameterIndex) {
        passiveParameterIndex = pPassiveParameterIndex;
    }

    @XmlAttribute
    public int getPassiveParameterIndex() {
        return passiveParameterIndex;
    }

    public void setPropulsionPlantConfiguration(short pPropulsionPlantConfiguration) {
        propulsionPlantConfiguration = pPropulsionPlantConfiguration;
    }

    @XmlAttribute
    public short getPropulsionPlantConfiguration() {
        return propulsionPlantConfiguration;
    }

    @XmlAttribute
    public short getNumberOfShafts() {
        return (short) shaftRPMs.size();
    }

    /** Note that setting this value will not change the marshalled value. The list whose length this describes is used for that purpose.
     * The getnumberOfShafts method will also be based on the actual list length rather than this value.
     * The method is simply here for java bean completeness.
     * @param pNumberOfShafts
     */
    public void setNumberOfShafts(short pNumberOfShafts) {
        numberOfShafts = pNumberOfShafts;
    }

    @XmlAttribute
    public short getNumberOfAPAs() {
        return (short) apaData.size();
    }

    /** Note that setting this value will not change the marshalled value. The list whose length this describes is used for that purpose.
     * The getnumberOfAPAs method will also be based on the actual list length rather than this value.
     * The method is simply here for java bean completeness.
     * @param pNumberOfAPAs
     */
    public void setNumberOfAPAs(short pNumberOfAPAs) {
        numberOfAPAs = pNumberOfAPAs;
    }

    @XmlAttribute
    public short getNumberOfUAEmitterSystems() {
        return (short) emitterSystems.size();
    }

    /** Note that setting this value will not change the marshalled value. The list whose length this describes is used for that purpose.
     * The getnumberOfUAEmitterSystems method will also be based on the actual list length rather than this value.
     * The method is simply here for java bean completeness.
     * @param pNumberOfUAEmitterSystems
     */
    public void setNumberOfUAEmitterSystems(short pNumberOfUAEmitterSystems) {
        numberOfUAEmitterSystems = pNumberOfUAEmitterSystems;
    }

    public void setShaftRPMs(List<ShaftRPMs> pShaftRPMs) {
        shaftRPMs = pShaftRPMs;
    }

    @XmlElementWrapper(name = "shaftRPMsList")
    public List<ShaftRPMs> getShaftRPMs() {
        return shaftRPMs;
    }

    public void setApaData(List<ApaData> pApaData) {
        apaData = pApaData;
    }

    @XmlElementWrapper(name = "apaDataList")
    public List<ApaData> getApaData() {
        return apaData;
    }

    public void setEmitterSystems(List<AcousticEmitterSystemData> pEmitterSystems) {
        emitterSystems = pEmitterSystems;
    }

    @XmlElementWrapper(name = "emitterSystemsList")
    public List<AcousticEmitterSystemData> getEmitterSystems() {
        return emitterSystems;
    }

    @Override
    public void marshal(DataOutputStream dos) {
        super.marshal(dos);
        try {
            emittingEntityID.marshal(dos);
            eventID.marshal(dos);
            dos.writeByte(stateChangeIndicator);
            dos.writeByte(pad);
            dos.writeShort((short) passiveParameterIndex);
            dos.writeByte((byte) propulsionPlantConfiguration);
            dos.writeByte((byte) shaftRPMs.size());
            dos.writeByte((byte) apaData.size());
            dos.writeByte((byte) emitterSystems.size());

            for (int idx = 0; idx < shaftRPMs.size(); idx++) {
                ShaftRPMs aShaftRPMs = shaftRPMs.get(idx);
                aShaftRPMs.marshal(dos);
            } // end of list marshalling


            for (int idx = 0; idx < apaData.size(); idx++) {
                ApaData aApaData = apaData.get(idx);
                aApaData.marshal(dos);
            } // end of list marshalling


            for (int idx = 0; idx < emitterSystems.size(); idx++) {
                AcousticEmitterSystemData aAcousticEmitterSystemData = emitterSystems.get(idx);
                aAcousticEmitterSystemData.marshal(dos);
            } // end of list marshalling

        } // end try
        catch (Exception e) {
            System.out.println(e);
        }
    } // end of marshal method

    @Override
    public void unmarshal(DataInputStream dis) {
        super.unmarshal(dis);

        try {
            emittingEntityID.unmarshal(dis);
            eventID.unmarshal(dis);
            stateChangeIndicator = dis.readByte();
            pad = dis.readByte();
            passiveParameterIndex = dis.readUnsignedShort();
            propulsionPlantConfiguration = (short) dis.readUnsignedByte();
            numberOfShafts = (short) dis.readUnsignedByte();
            numberOfAPAs = (short) dis.readUnsignedByte();
            numberOfUAEmitterSystems = (short) dis.readUnsignedByte();
            for (int idx = 0; idx < numberOfShafts; idx++) {
                ShaftRPMs anX = new ShaftRPMs();
                anX.unmarshal(dis);
                shaftRPMs.add(anX);
            }

            for (int idx = 0; idx < numberOfAPAs; idx++) {
                ApaData anX = new ApaData();
                anX.unmarshal(dis);
                apaData.add(anX);
            }

            for (int idx = 0; idx < numberOfUAEmitterSystems; idx++) {
                AcousticEmitterSystemData anX = new AcousticEmitterSystemData();
                anX.unmarshal(dis);
                emitterSystems.add(anX);
            }

        } // end try
        catch (Exception e) {
            System.out.println(e);
        }
    } // end of unmarshal method

    /**
     * Packs a Pdu into the ByteBuffer.
     * @throws java.nio.BufferOverflowException if buff is too small
     * @throws java.nio.ReadOnlyBufferException if buff is read only
     * @see java.nio.ByteBuffer
     * @param buff The ByteBuffer at the position to begin writing
     * @since ??
     */
    @Override
    public void marshal(java.nio.ByteBuffer buff) {
        super.marshal(buff);
        emittingEntityID.marshal(buff);
        eventID.marshal(buff);
        buff.put(stateChangeIndicator);
        buff.put(pad);
        buff.putShort((short) passiveParameterIndex);
        buff.put((byte) propulsionPlantConfiguration);
        buff.put((byte) shaftRPMs.size());
        buff.put((byte) apaData.size());
        buff.put((byte) emitterSystems.size());

        for (int idx = 0; idx < shaftRPMs.size(); idx++) {
            ShaftRPMs aShaftRPMs = shaftRPMs.get(idx);
            aShaftRPMs.marshal(buff);
        } // end of list marshalling


        for (int idx = 0; idx < apaData.size(); idx++) {
            ApaData aApaData = apaData.get(idx);
            aApaData.marshal(buff);
        } // end of list marshalling


        for (int idx = 0; idx < emitterSystems.size(); idx++) {
            AcousticEmitterSystemData aAcousticEmitterSystemData = emitterSystems.get(idx);
            aAcousticEmitterSystemData.marshal(buff);
        } // end of list marshalling

    } // end of marshal method

    /**
     * Unpacks a Pdu from the underlying data.
     * @throws java.nio.BufferUnderflowException if buff is too small
     * @see java.nio.ByteBuffer
     * @param buff The ByteBuffer at the position to begin reading
     * @since ??
     */
    @Override
    public void unmarshal(java.nio.ByteBuffer buff) {
        super.unmarshal(buff);

        emittingEntityID.unmarshal(buff);
        eventID.unmarshal(buff);
        stateChangeIndicator = buff.get();
        pad = buff.get();
        passiveParameterIndex = (buff.getShort() & 0xFFFF);
        propulsionPlantConfiguration = (short) (buff.get() & 0xFF);
        numberOfShafts = (short) (buff.get() & 0xFF);
        numberOfAPAs = (short) (buff.get() & 0xFF);
        numberOfUAEmitterSystems = (short) (buff.get() & 0xFF);
        for (int idx = 0; idx < numberOfShafts; idx++) {
            ShaftRPMs anX = new ShaftRPMs();
            anX.unmarshal(buff);
            shaftRPMs.add(anX);
        }

        for (int idx = 0; idx < numberOfAPAs; idx++) {
            ApaData anX = new ApaData();
            anX.unmarshal(buff);
            apaData.add(anX);
        }

        for (int idx = 0; idx < numberOfUAEmitterSystems; idx++) {
            AcousticEmitterSystemData anX = new AcousticEmitterSystemData();
            anX.unmarshal(buff);
            emitterSystems.add(anX);
        }

    } // end of unmarshal method

    /**
     * The equals method doesn't always work--mostly on on classes that consist only of primitives. Be careful.
     * @param rhs
     * @return
     */
    public boolean equals(UaPdu rhs) {
        boolean ivarsEqual = true;

        if (rhs.getClass() != this.getClass()) {
            return false;
        }

        if (!(emittingEntityID.equals(rhs.emittingEntityID))) {
            ivarsEqual = false;
        }
        if (!(eventID.equals(rhs.eventID))) {
            ivarsEqual = false;
        }
        if (!(stateChangeIndicator == rhs.stateChangeIndicator)) {
            ivarsEqual = false;
        }
        if (!(pad == rhs.pad)) {
            ivarsEqual = false;
        }
        if (!(passiveParameterIndex == rhs.passiveParameterIndex)) {
            ivarsEqual = false;
        }
        if (!(propulsionPlantConfiguration == rhs.propulsionPlantConfiguration)) {
            ivarsEqual = false;
        }
        if (!(numberOfShafts == rhs.numberOfShafts)) {
            ivarsEqual = false;
        }
        if (!(numberOfAPAs == rhs.numberOfAPAs)) {
            ivarsEqual = false;
        }
        if (!(numberOfUAEmitterSystems == rhs.numberOfUAEmitterSystems)) {
            ivarsEqual = false;
        }

        for (int idx = 0; idx < shaftRPMs.size(); idx++) {
            ShaftRPMs x = shaftRPMs.get(idx);
            if (!(shaftRPMs.get(idx).equals(rhs.shaftRPMs.get(idx)))) {
                ivarsEqual = false;
            }
        }


        for (int idx = 0; idx < apaData.size(); idx++) {
            ApaData x = apaData.get(idx);
            if (!(x.equals(rhs.apaData.get(idx)))) {
                ivarsEqual = false;
            }
        }


        for (int idx = 0; idx < emitterSystems.size(); idx++) {
            AcousticEmitterSystemData x = emitterSystems.get(idx);
            if (!(x.equals(rhs.emitterSystems.get(idx)))) {
                ivarsEqual = false;
            }
        }

        return ivarsEqual;
    }
} // end of class
